### ISSUE TITLE
Validation fails with --no-environment flag if environment variables are not defined

### ISSUE THREAD

**Issue by tie** [vectordotdev/vector#21270]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

After #20422, validating Vector configuration that uses environment variables fails even with `--no-environment` flag. A temporary workaround is `${VAR-}` (default to an empty string), but then there is no error if the variable is actually undefined at runtime.

### Configuration

```text
[sinks.push-metrics]
endpoint = "<REDACTED>"
inputs = ["metrics"]
type = "prometheus_remote_write"
[sinks.push-metrics.auth]
password = "$METRICS_PASSWORD"
strategy = "basic"
user = "<REDACTED>"

[sinks.push-metrics.healthcheck]
enabled = false

[sources]
[sources.host-metrics]
type = "host_metrics"

[transforms.metrics]
inputs = ["host-metrics"]
source = ".tags.env = \"production\"\n.tags.host = \"<REDACTED>\"\n"
type = "remap"
```

### Version

vector 0.38.0 (aarch64-unknown-linux-gnu)

### Debug Output

```text
2024-09-11T04:57:21.709836Z DEBUG vector::app: Internal log rate limit configured. internal_log_rate_secs=10
2024-09-11T04:57:21.711457Z  INFO vector::app: Log level is enabled. level="trace"
2024-09-11T04:57:21.711616Z DEBUG vector::app: messaged="Building runtime." worker_threads=16
2024-09-11T04:57:21.713223Z TRACE mio::poll: registering event source with poller: token=Token(1), interests=READABLE    
Failed to load ["/nix/store/imjx3c2bmvglkw5cjnc3js5c7yscr4c5-vector.toml"]
--------------------------------------------------------------------------
x Missing environment variable in config. name = "METRICS_PASSWORD"
```

**Comment by tie** [comment]
> I think there may be some misunderstanding of what --no-environment does. That actually just doesn't run the **health checks**, rather than not requiring environment variables to be defined.

Aren’t health checks controlled by a separate [`--skip-healthchecks` flag](https://vector.dev/docs/administration/validating/#skipping-health-checks) (that `--no-environment` implies)?

From https://vector.dev/docs/administration/validating/#environment-checks:
>1. All components have the pre-requisites to run, e.g. data directories exist and are writable.
>
>[…]
>
>These environment checks can be disabled using the [--no-environment](https://vector.dev/docs/reference/cli/#validate-no-environment) flag: […]

I’d expect environment variables to be part of the pre-requisites to run. If that’s not the case, I think this should be explicitly documented given that the flag name is rather misleading.

>If the environment variables are not available when you are running `vector validate` you can provide placeholder values like: `METRICS_PASSWORD=foo vector validate`. Does that make sense?

In this particular case, environment file is generated from secrets, so we might as well generate a JSON file and introduce `file` secrets backend instead of using environment variables (and reload configuration instead of restarting the service on secret changes). This looks trivial to implement, I’ll create an issue/PR with the implementation.

**Comment by jszwedko** [comment]
Yeah, I definitely agree that the flag name encourages the assumption that environment variables won't be required to be defined. The naming could be better.

Unfortunately, we can't not require environment variables, currently, because otherwise the config cannot be validated without them as they can appear anywhere in the config (and even template in whole components). There has been discussion about restricting env vars to only appearing in certain places (e.g. values for options) at which point I think it would be possible to validate config without them.

`--no-environment` should skip health checks _and_ checking for the data directory. `--skip-healthchecks` can be used to skip just health checks where the data directory will still be validated.

A `file` secrets back-end would be more than welcome! Thanks for opening the PR!
